jgrapht-core/src/main/java/org/jgrapht/alg/scoring/ClusteringCoefficient.java

```diff
@@ -0,0 +1,215 @@
+    }
+
+    /**
+     * Computes the average clustering coefficient. The average clustering coefficient $\={C}$ is defined as
+     * $\={C} = \frac{\sum_{i=1}^{n} C_i}{n}$ where $n$ is the number of vertices.
+     *
+     * Note: the average is $0$ if the graph is empty
+     *
+     * @return the average clustering coefficient
+     */
+    public double getAverageClusteringCoefficient(){
+        if (graph.vertexSet().size() == 0)
+            return 0;
+
+        if (!computedAverage){
+            computeScoreMap();
+            computedAverage = true;
+            averageClusteringCoefficient = 0;
+
+            for (double value: scores.values())
+                averageClusteringCoefficient += value;
+
+            averageClusteringCoefficient /= graph.vertexSet().size();
+        }
+
+        return averageClusteringCoefficient;
+    }
+
+    private void computeGlobalClusteringCoefficient(){
+        NeighborCache<V, E> neighborCache = new NeighborCache<>(graph);
+        computed = true;
+        double numberTriplets = 0;
+
+        for (V v: graph.vertexSet()){
+            if (graph.getType().isUndirected()){
+                numberTriplets += 1.0 * graph.degreeOf(v) * (graph.degreeOf(v) - 1) / 2;
+            }
+            else{
+                numberTriplets += 1.0 * neighborCache.predecessorsOf(v).size() * neighborCache.successorsOf(v).size();
+            }
+        }
+
+        globalClusteringCoefficient = 3 * GraphMetrics.getNumberOfTriangles(graph) / numberTriplets;
+    }
+
+    private void computeScoreMap(){
+        if (scores != null)
+            return;
+
+        scores = new HashMap<>(graph.vertexSet().size());
+
+        for (V v: graph.vertexSet()){
+            NeighborCache<V, E> neighborCache = new NeighborCache<>(graph);
+            Set<V> neighbourhood = neighborCache.neighborsOf(v);
+
+            final double k = neighbourhood.size();
+            double numberTriplets = 0;
+
+            for (V p: neighbourhood)
+                for (V q: neighbourhood)
+                    if (graph.containsEdge(p, q))
+                        numberTriplets++;
+
+            if (k <= 1)
+                scores.put(v, 0.0);
+            else
+                scores.put(v, numberTriplets / (k * (k - 1)));
+        }
+    }
+
+    /**
+     * Get a map with the local clustering coefficients of all vertices
+     *
+     * @return a map with all local clustering coefficients
+     */
+    @Override
+    public Map<V, Double> getScores() {
+        computeScoreMap();
+        return Collections.unmodifiableMap(scores);
+    }
+
+    /**
+     * Get a vertex's local clustering coefficient
+     *
+     * @param v the vertex
+     * @return the local clustering coefficient
+     */
+    @Override
+    public Double getVertexScore(V v) {
+        if (!graph.containsVertex(v)) {
+            throw new IllegalArgumentException("Cannot return score of unknown vertex");
+        }
+
+        computeScoreMap();
```

Comment:
Why do we compute the complete score map when a single vertex score is requested? It is not always the case that we would want the average/global coefficient.

Reply:
You are right. There's no need to computed the whole map. I've changed it to do the computation per vertex when needed.